db/db_impl/db_impl_write.cc

```diff
@@ -823,6 +840,17 @@ void DBImpl::WriteStatusCheck(const Status& status) {
   }
 }
 
+void DBImpl::IOStatusCheck(const IOStatus& io_status) {
```

Comment:
It could be in another PR, but I believe that ```IOStatusCheck()``` and ```WriteStatusCheck()``` should propagate the status returned by ```SetBGError``` to ```w.status```.